Adjust cursor animator properties

mwyrzykowski opened the pull request:
#### cebe6cd67d9d949d227afe615b2965806c96b71a
<pre>
Adjust cursor animator properties
<a href="https://bugs.webkit.org/show_bug.cgi?id=255399">https://bugs.webkit.org/show_bug.cgi?id=255399</a>
&lt;radar://107662722&gt;

Reviewed by Aditya Keerthi.

This patch allows suspending the blink state of the
cursor from the UI process, moves logic from an
static internal function to a virtual function,
and corrects an issue where the caret does not
start animating after a transition.

This also removes CaretAnimator::addLine which is no longer used.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::CaretBase::paintCaret const):
(WebCore::FrameSelection::caretAnimatorInvalidated):
Ensure animator starts animating after a caret transition.

(WebCore::FrameSelection::updateAppearance):
With CaretAnimator&apos;s stop function virtual, send the reason
for why the stop was issued.

* Source/WebCore/platform/CaretAnimator.cpp:
(WebCore::CaretAnimator::adjustedClipRect const):
Allow specific animator&apos;s to override the clip rectangle.

* Source/WebCore/platform/CaretAnimator.h:
(WebCore::CaretAnimator::stop):
Allow for a reason to be used in determining stop behavior.

(WebCore::CaretAnimator::addLine const): Deleted.
This is no longer needed.

* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::paint):
(WebCore::RenderBlock::paintObject):
Revert previously added changes.

(WebCore::renderCaretInsideContentsClip): Deleted.
This is no longer used, instead CaretAnimator has a virtual member
function to control the clip rectangle.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setCaretBlinkingSuspended):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setCaretBlinkingSuspended):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
Allow blink to be suspended from the UI process.

Canonical link: <a href="https://commits.webkit.org/263083@main">https://commits.webkit.org/263083@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/553690682ba820e9fa60ead869857dc58c54b227

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3560 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3607 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3746 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4984 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3835 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3539 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3700 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3648 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3093 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3603 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3855 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/3221 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4806 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1362 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3183 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/3084 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3162 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/3242 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4571 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3619 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2922 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3160 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3181 "Passed tests") | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/861 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3187 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3438 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -36,6 +36,11 @@ enum class CaretAnimatorType : uint8_t {
+enum class CaretAnimatorStopReason : uint8_t {
+    Default,
+    CaretRectChanged,
+};
```
Slightly strange to see this before the forward declarations. Would you mind moving these enums below line 50?

pxlcoder commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -70,7 +75,7 @@ class CaretAnimator {
-    void stop()
+    virtual void stop(CaretAnimatorStopReason = CaretAnimatorStopReason::Default)
     {
         if (!m_isActive)
```
Should probably move this into the implementation file.

pxlcoder commented on Source/WebCore/platform/CaretAnimator.h:
```diff
@@ -91,7 +96,9 @@ class CaretAnimator {
     PresentationProperties presentationProperties() const { return m_presentationProperties; }
     virtual void paint(const Node&, GraphicsContext&, const FloatRect&, const Color&, const LayoutPoint&) const;
     virtual LayoutRect repaintCaretRectForLocalRect(LayoutRect) const;
+#if defined(__has_include) && __has_include(<WebKitAdditions/RenderBlockAdditions.h>)
```
Probably want a `FIXME` + bug, so we can remember to remove this `#if`?

mwyrzykowski replied:
Good idea, I added a FIXME with a bug

mwyrzykowski committed:
Adjust cursor animator properties
https://bugs.webkit.org/show_bug.cgi?id=255399
<radar://107662722>

Reviewed by Aditya Keerthi.

This patch allows suspending the blink state of the
cursor from the UI process, moves logic from an
static internal function to a virtual function,
and corrects an issue where the caret does not
start animating after a transition.

This also removes CaretAnimator::addLine which is no longer used.

* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::CaretBase::paintCaret const):
(WebCore::FrameSelection::caretAnimatorInvalidated):
Ensure animator starts animating after a caret transition.

(WebCore::FrameSelection::updateAppearance):
With CaretAnimator's stop function virtual, send the reason
for why the stop was issued.

* Source/WebCore/platform/CaretAnimator.cpp:
(WebCore::CaretAnimator::adjustedClipRect const):
Allow specific animator's to override the clip rectangle.

* Source/WebCore/platform/CaretAnimator.h:
(WebCore::CaretAnimator::stop):
Allow for a reason to be used in determining stop behavior.

(WebCore::CaretAnimator::addLine const): Deleted.
This is no longer needed.

* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::paint):
(WebCore::RenderBlock::paintObject):
Revert previously added changes.

(WebCore::renderCaretInsideContentsClip): Deleted.
This is no longer used, instead CaretAnimator has a virtual member
function to control the clip rectangle.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::setCaretBlinkingSuspended):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::setCaretBlinkingSuspended):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
Allow blink to be suspended from the UI process.

Canonical link: https://commits.webkit.org/263083@main
